hw/ip/otbn/dv/tracer/cpp/log_trace_listener.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cassert>
+#include <iomanip>
+#include <ios>
+#include <sstream>
+#include <string>
+
+#include "log_trace_listener.h"
+
+LogTraceListener::LogTraceListener(const std::string &log_filename)
+    : trace_log(log_filename, std::fstream::out) {
+  if (!trace_log.is_open()) {
+    std::ostringstream oss;
+    oss << "Could not open log file: " << log_filename;
+    throw std::runtime_error(oss.str());
+  }
+}
+
+void LogTraceListener::AcceptTraceString(const std::string &trace,
+                                         unsigned int cycle_count) {
+  assert(trace_log.is_open());
+
+  // Split the trace up into a vector of strings, one per line
+  auto trace_lines = SplitTraceLines(trace);
+  bool first_line = true;
+
+  // Write out the lines from the trace
+  for (auto &line : trace_lines) {
+    if (first_line) {
+      // Special handling for the first line
+      assert(line.size() > 1);
+
+      // It is expected the first line of any trace output is an 'E' or 'S' line
+      // (instruction execute or instruction stall)
+      bool is_e_or_s_line = line[0] == 'E' || line[0] == 'S';
+
+      // Output the beginning of the first line adding a cycle count. If this
+      // isn't an 'E' or 'S' line use a special '!' to indicate the lack of one.
+      std::ios old_state(nullptr);
+      old_state.copyfmt(trace_log);
```

Comment:
This is fine but, since you control `trace_log`, you could just assume that the fill was ' ' before.

Reply:
True, I guess I prefer this way as it doesn't mean we have to remember '' is meant to be the default fill character for trace_log output.